platform: make C++ allocation wrappers log the correct caller address

TeroJaasko opened the pull request:
## Description

The C++ "operator new" and "operator delete" (and their array
variants) were logging the the caller address wrong. In practice
if one used "operator new", the logged caller address pointed
to mbed_retarget.cpp, not to the client. Fix this by exposing
the alloc wrappers to the the retarget.

Note: this fixes only the ARMCC variants, as the GCC ones have
different different API and implementation.

Related mbed-os issue: https://github.com/ARMmbed/mbed-os/issues/5221

## Status

**IN DEVELOPMENT**, GCC should also be fixed.

## Migrations

NO

## Related PRs


## Todos

- [x] Tests
- [ ] Documentation

## Deploy notes

Note: the GCC target is equally broken, but this PR is not fixing that.

## Steps to test or reproduce
Test code
```
class test_class
{
public:
    uint32_t member;
};

static void do_the_test()
{
    void *mem_malloc = malloc(100);
    void *mem_calloc = calloc(60, 2);
    test_class* cpp_new_obj = new test_class;
    uint8_t* cpp_new_array = new uint8_t[200];
    delete cpp_new_obj;
    delete[] cpp_new_array;
    free(mem_calloc);
    free(mem_malloc);
}
```

0xc0170 commented:
> Note: the GCC target is equally broken, but this PR is not fixing that.

@TeroJaasko Can we fix also GCC in this patch, or are there any blockers ?

> this fixes only the ARMCC variants

I can see also IAR is fixed. In the code please can you change `TOOLCHAIN_ARM` to the symbol that ARMCC provides?

TeroJaasko commented:
@0xc0170: GCC variant has those "struct _reent" parameters in the wrappers, eg. "__real__malloc_r(struct _reent * r, size_t size)". I have no idea where to get that from C++ side. Any pointers where to look? 

The IAR build does not seem to get the caller address at all, as the MBED_CALLER_ADDRESS macro is defined as NULL, see https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_toolchain.h#L314. Is there a way to get the caller address on IAR? Or should I just remove the 
IAR part from ifdef?

Regarding TOOLCHAIN_ARM - which is the correct one to use instead of it? On the mbed_retarget.cpp there are __ARMCC_VERSION and __CC_ARM used in addition of TOOLCHAIN_ARM.

0xc0170 commented:
> @0xc0170: GCC variant has those "struct _reent" parameters in the wrappers, eg. "__real__malloc_r(struct _reent * r, size_t size)". I have no idea where to get that from C++ side. Any pointers where to look?

Don't know from top of my head, @pan- any pointers?

> The IAR build does not seem to get the caller address at all, as the MBED_CALLER_ADDRESS macro is defined as NULL, see https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_toolchain.h#L314. Is there a way to get the caller address on IAR? Or should I just remove the
IAR part from ifdef?

Looks correct, not sure if this changes with IAR 8 update, I could not find it.

> Regarding TOOLCHAIN_ARM - which is the correct one to use instead of it? On the mbed_retarget.cpp there are __ARMCC_VERSION and __CC_ARM used in addition of TOOLCHAIN_ARM.

`__CC_ARM` should be sufficient

pan- commented:
> I have no idea where to get that from C++ side. Any pointers where to look?

You may include `<reent.h>` and use the macro [`_REENT`](https://github.com/bminor/newlib/blob/master/newlib/libc/include/sys/reent.h#L782-L789) but there is no guarantee this will work long term.

theotherjimmy commented:
@TeroJaasko Could you fix this in all 3 compilers before we merge it?

TeroJaasko commented:
@pan- : thanks for guidance, that seems to work. 
@theotherjimmy: Added the GCC support and changed the TOOLCHAIN_ARM to __CC_ARM. 

IAR 8.x version seem to mostly work also if I use "#define MBED_CALLER_ADDR() ((void *)__get_LR())". I don't have IAR 7 installed, so if possible, could you take the ball on the IAR support?

adbridge commented:
@TeroJaasko Could you please rebase this to remove the conflict ?

TeroJaasko commented:
@adbridge : done. At same time, removed also the pointless "if (ptr)"'s from the free wrappers as they were removed from the original versions.

0xc0170 commented:
I restarted jenkins CI (network failure there)

@TeroJaasko what is outstanding here?

> IAR 8.x version seem to mostly work also if I use "#define MBED_CALLER_ADDR() ((void *)__get_LR())". I don't have IAR 7 installed, so if possible, could you take the ball on the IAR support?

IAR support (version 7.x) ?

TeroJaasko commented:
As far as I know, the IAR target has never been able to print the caller address at all, just zeros. If you want to fix also IAR as part of this PR, please do so.

0xc0170 commented:
Summarize: ARMCC and GCC fixed, IAR does not provide caller address thus no change there. 

This fix is ready for review

TeroJaasko commented:
CI build pointed out that I had broken the non-tracing IAR build. Fixed it and force pushed.

0xc0170 commented:
/morph build

cmonr commented:
@TeroJaasko Did you intend on adding any additional documentation?

TeroJaasko commented:
@cmonr : I have nothing to add here.
